daemons/fenced/pacemaker-fenced.c

```diff
@@ -578,7 +579,44 @@ our_node_allowed_for(pe_resource_t *rsc)
 }
 
 static void
-watchdog_device_update(xmlNode *cib)
+watchdog_device_update(void)
+{
+    if (stonith_watchdog_timeout_ms > 0) {
+        if (!g_hash_table_lookup(device_list, STONITH_WATCHDOG_ID) &&
+            !stonith_watchdog_targets) {
+            /* getting here watchdog-fencing enabled, no device there yet
+               and reason isn't stonith_watchdog_targets preventing that
+             */
+            int rc;
+            xmlNode *xml;
+
+            xml = create_device_registration_xml(
+                    STONITH_WATCHDOG_ID,
+                    st_namespace_internal,
+                    STONITH_WATCHDOG_AGENT,
+                    NULL, /* stonith_device_register will add our
+                             own name as PCMK_STONITH_HOST_LIST param
+                             so we can skip that here
+                           */
+                    NULL);
+            rc = stonith_device_register(xml, NULL, FALSE);
+            free_xml(xml);
+            if (rc != pcmk_ok) {
+                crm_crit("Cannot register watchdog pseudo fence agent");
+                crm_exit(CRM_EX_FATAL);
+            }
+        }
+
+    } else {
+        /* be silent if no device - todo parameter to stonith_device_remove */
+        if (g_hash_table_lookup(device_list, STONITH_WATCHDOG_ID)) {
+            stonith_device_remove(STONITH_WATCHDOG_ID, TRUE);
+        }
```

Comment:
Now I'm second-guessing myself. Above, we call stonith_device_register(..., FALSE). So, it can be CIB-registered by the user, or API-registered internally. I guess we need to unregister it from both here? Maybe stonith_device_unregister() should take flags instead of boolean so either or both CIB/API can be cleared? Then you could easily add a flag for "be silent if no device" too ...

Reply:
Guess changing stonith_device_register / remove to flags instead of this boolean is a good idea.
But it is should be done separately I guess.
For now I guess doing stonith_device_register(..., TRUE) here is fine as well.
Probably even more reflects reality - not added as a fencing-resource configured in the cib but via a property coming from the cib.

Reply:
The problem is, if the user configured fence_watchdog in the CIB, then it will naturally get CIB-registered, but above this where we create an implicit registration with create_device_registration_xml(), we register it with stonith_device_register(xml, NULL, TRUE) (where TRUE means API-registered). I suppose we could just change that to FALSE, so it's always CIB-registered.

Reply:
which is what I had just done in my last update ... at least I think so ...
I think the TRUE means via CIB and is consistent with remove.